File: src/lib/events.ts
// Typed wrappers around Tauri events emitted by the Rust core.
//
// Event channel names mirror `src-tauri/src/events.rs`. Payload shapes are kept
// in sync with the Rust `serde` structs. Subscribe via the helpers below so the
// channel-name strings live in exactly one place.
import { listen, type UnlistenFn } from "@tauri-apps/api/event";

/** Whether a stats snapshot is live or a marked-stale cache. */
export type StatsSource = { kind: "live" } | { kind: "stale"; age_secs: number };

/** Normalized miner telemetry (mirrors Rust `MinerStats`). */
export interface MinerStats {
  hashrate_hs: number;
  hashrate_avg: [number, number, number];
  accepted: number;
  rejected: number;
  best_share_diff: number;
  uptime_secs: number;
  connected: boolean;
  current_job_diff: number | null;
  source: StatsSource;
}

export type MinerState =
  | "idle"
  | "starting"
  | "running"
  | "reconnecting"
  | "crashed"
  | "stopped";

export interface StatsEvent {
  miner_id: string;
  stats: MinerStats;
}
export interface StatusEvent {
  miner_id: string;
  state: MinerState;
  detail: string | null;
}
export interface LogEvent {
  miner_id: string;
  line: string;
}
export interface BlockFoundEvent {
  miner_id: string;
  coin: string;
  share_diff: number;
  network_diff: number;
}

export const Channels = {
  stats: "miner://stats",
  status: "miner://status",
  log: "miner://log",
  blockFound: "miner://block-found",
  difficulty: "network://difficulty",
} as const;

/** Generic typed subscription helper. */
export function on<T>(
  channel: string,
  handler: (payload: T) => void,
): Promise<UnlistenFn> {
  return listen<T>(channel, (event) => handler(event.payload));
}

export const onStats = (h: (e: StatsEvent) => void) => on<StatsEvent>(Channels.stats, h);
export const onStatus = (h: (e: StatusEvent) => void) => on<StatusEvent>(Channels.status, h);
export const onLog = (h: (e: LogEvent) => void) => on<LogEvent>(Channels.log, h);
export const onBlockFound = (h: (e: BlockFoundEvent) => void) =>
  on<BlockFoundEvent>(Channels.blockFound, h);

File: src-tauri/src/events.rs
//! Typed event payloads and channel names emitted to the frontend.
//!
//! Channel-name constants are the single source of truth shared (by convention)
//! with `src/lib/events.ts`. Payload structs derive `Serialize` so the wire
//! shape stays in sync with the TypeScript types.

use serde::Serialize;

use crate::miner::MinerStats;

/// Event channel names. Keep in lockstep with `Channels` in `src/lib/events.ts`.
pub mod channel {
    pub const STATS: &str = "miner://stats";
    pub const STATUS: &str = "miner://status";
    pub const LOG: &str = "miner://log";
    pub const BLOCK_FOUND: &str = "miner://block-found";
    pub const DIFFICULTY: &str = "network://difficulty";
}

/// A telemetry snapshot for one running miner instance.
#[derive(Debug, Clone, Serialize)]
pub struct StatsEvent {
    pub miner_id: String,
    pub stats: MinerStats,
}

/// A single line of miner stdout/stderr forwarded to the UI log viewer.
#[derive(Debug, Clone, Serialize)]
pub struct LogEvent {
    pub miner_id: String,
    pub line: String,
}

/// Lifecycle state of a managed miner instance.
#[derive(Debug, Clone, Copy, Serialize, PartialEq, Eq)]
#[serde(rename_all = "lowercase")]
pub enum MinerState {
    Idle,
    Starting,
    Running,
    Reconnecting,
    Crashed,
    Stopped,
}

/// Status transition for a miner instance.
#[derive(Debug, Clone, Serialize)]
pub struct StatusEvent {
    pub miner_id: String,
    pub state: MinerState,
    pub detail: Option<String>,
}

/// A found block (best share reached network difficulty) — the rare win.
#[derive(Debug, Clone, Serialize)]
pub struct BlockFoundEvent {
    pub miner_id: String,
    pub coin: String,
    pub share_diff: f64,
    pub network_diff: f64,
}
